---
title: "Genome annotation (part 3)"
description: In this lab you we continue putting together the building blocks of a genome annotation suite by writing a function to find open reading frames (ORFs).
section: Exercises
data: data/chr01-truncated.fsa
---

# Setup

We're going to put the pieces of the last two exercises together with some new higher level functions to actually start annotating a chromosome.  Include your refined implementations for the `transcribe`, `translate` and `findORF` functions here:

``` {r grade=TRUE}
# Transcibes DNA sequence into RNA equivalent
#
# dna   A character string containing DNA sequence
#
# value A character string containing RNA sequence
#
transcribe <- function(dna) {
  gsub( pattern     = "T"
      , replacement = "U"
      , x           = toupper(dna)
      , fixed       = TRUE
      )
}

# Translates RNA sequence into amino acid sequence
#
# rna   A character string containing RNA sequence
#
# value A character string containing amino sequence
#
translate <- function(rna) {

  codonTable  <- read.table( "data/codons.txt"
                           , header            = TRUE
                           , row.names         = "codon"
                           , stringsAsFactors  = FALSE
                           )

  codons      <- sapply( seq( from = 1
                            , to   = nchar(rna) - (nchar(rna) %% 3)
                            , by   = 3 
                            )
                       , function(n) {
                            substr(rna, start = n, stop = n + 2)
                         }
                       )
  
  paste( codonTable[codons, "aminoAcid"]
       , collapse = ""
       )
}

# Returns a data.frame describing the ORFs found in the given aminoacid sequence
#
# aminoacids  A character string containing aminoacid sequence
# cutoff      An integer giving the minimum aminoacid length of valid ORFs
#
# value       A data.frame, coordinates refer to amino acid position
#
findORF <- function(aminoacids, cutoff = 100) {

  matches <- gregexpr(pattern = "M(.*?)X", text = aminoacids)
  starts  <- as.vector(matches[[1]])
  lengths <- attr(matches[[1]], "match.length")
  
  df <- data.frame( aa.start   = starts
                  , aa.length  = lengths
                  , aa.seq     = substring( aminoacids
                                          , first = starts
                                          , last  = starts + lengths - 1
                                          )
                  )
  
  df[df$aa.length >= cutoff,]
}
```

# Step 3:  Annotate a complete chromosome

## Overview

Now we're ready to think about the structure of our function which will annotate all of the possible ORFs in the sequence for a chromosome.  Here's the template we'll use for our big annotation function:

``` {r}
# Finds all open reading frames in the chromosome sequence in a FASTA file
#
# fastaFile   A string containing the name of a chromosome file in FASTA format
#
# value       A data.frame.  Coordinates are relative to DNA sequence in input.
#
annotateChromosome <- function(fastaFile) {


}
```

```{r grade=TRUE}
library(plyr)

# Finds all open reading frames in the chromosome sequence in a FASTA file
#
# fastaFile   A string containing the name of a chromosome file in FASTA format
#
# value       A data.frame.  Coordinates are relative to DNA sequence in input.
#
annotateChromosome <- function(fastaFile) {

  forwardStrand <- transcribe(loadFASTA(fastaFile))
  reverseStrand <- reverseComplement(forwardStrand)
  
  forwardFrames <- lapply( c(0, 1, 2)
                         , function(offset) { 
                              annotateFrame(forwardStrand, offset)
                           }  
                         )
  
  # for the reverse frames, we'll need to adjust the starting coordinate to be 
  # the end of the orf in the 5'->3' direction of the reference (forward)
  # strand.  Draw this out if that doesn't make sense...
  reverseFrames <- lapply( c(0, 1, 2)
                         , function(offset) { 
                             orfs <- annotateFrame(forwardStrand, offset)
                             orfs$start <- nchar(reverseStrand) - orfs$start + 1
                             orfs
                           }  
                         )

  # Since we're introducing plyr in today's lab, I'll use it here for 
  # illustration.  Note calling `c` on two lists just concatenates the elements 
  # of each list (in this case data.frames with orf annotations) into a new 
  # list.
  ldply(c(forwardFrames, reverseFrames))
  
}
```

We code write all of the code we need to accomplish the goals of `annotateChromosome` in one ginormous function body.  But it's always a good idea to avoid doing this: as a rule of thumb if a function has grown beyond a dozen or so lines, you might want to break it up into sub-parts to make it easier to read and alter in the future.  So we'll also write a few helper functions to keep annotateChromosome from becoming bloated.

Here are the templates:

``` {r}
# Parses a sequence file in FASTA format
#
# fastaFile   A string giving the name of a fasta file to parse
#
# value       A string containing the parsed sequence
#
loadFASTA <- function(fastaFile) {


}
```

```{r grade=TRUE}
# Parses a sequence file in FASTA format
#
# fastaFile   A string giving the name of a fasta file to parse
#
# value       A string containing the parsed sequence
#
loadFASTA <- function(fastaFile) {

  # trim FASTA header line; concatenate rows of sequence.
  lines <- readLines(fastaFile)
  paste( lines[2:length(lines)]
       , collapse = ""
       , warn     = FALSE
       )

}
```

```{r}
# Annotates the ORFs found in a given reading frame of a dnaString
#
# dnaStrand   A string containing DNA sequence
# offset      The frame offset (0, 1, or 2)
#
# value       A data.frame with a `length` and `startPosition` column. 
#             Coordinates are relative to dnaStrand.
#
annotateFrame <- function(dnaStrand, offset) {


}
```

```{r grade=TRUE}
# Annotates the ORFs found in a given reading frame of a dnaString
#
# dnaStrand   A string containing DNA sequence
# offset      The frame offset (0, 1, or 2)
#
# value       A data.frame with a `length` and `start` column. 
#             Coordinates are relative to dnaStrand.
#
annotateFrame <- function(dnaStrand, offset) {

  searchFrame <- substr( dnaStrand
                       , start = 1 + offset
                       , stop  = nchar(dnaStrand)
                       )
  
  orfs        <- findORF(translate(searchFrame))
  
  orfs$length <- orfs$aa.length * 3
  orfs$start  <- (orfs$aa.start * 3) + offset
  
  orfs

}
```

```{r}
# Calculates the reverse complement of a dnaStrand
#
# dnaStrand   A string containing the forward DNA sequence
#
# value       A string containing the reverse complement to dnaStrand
#
reverseComplement <- function(dnaStrand) {


}
```

```{r grade=TRUE}
# Calculates the reverse complement of a dnaStrand
#
# dnaStrand   A string containing the forward DNA sequence
#
# value       A string containing the reverse complement to dnaStrand
#
reverseComplement <- function(dnaStrand) {

  forward <- c("A","T","G","C")
  reverse <- c("t","a","c","g")
  
  complement <- reverseString(dnaStrand)
  
  # see ?mapply; like sapply but takes multiple vector arguments.
  mapply( function(f, r) { 
    
            # the <<- operator assigns data into a variable in the parent
            # context;  in this case each time the next line is executed, it
            # will replace the value for the `complement` variable declared
            # above.
            complement <<- gsub(f, r, complement, fixed = TRUE)
            
          }
        , forward
        , reverse
        )
  
  toupper(complement)
}
```

```{r}
# Reverses the characters in a string
reverseString <- function(a) {
  
}
```

```{r grade=TRUE}
# Reverses the characters in a string
reverseString <- function(s) {
  paste(rev(substring(s, 1:nchar(s), 1:nchar(s))), collapse = "") 
}
```

I'm not going to walk you through exactly how to write the body of each of these functions.  Instead, I'll give some pointers for each in the sections that follow.

## reverseString and reverseComplement

To implement this function you'll probably want to start by filling in the `reverseString` utility function.  For that job you might want to use the `strsplit` function (using `split = ""` will split a string into it's characters).  Note that `strsplit` returns a `list` with one element for each string in the character vector you give it.

For example:
```{r}
strsplit(   "abcd"         , split = "")
strsplit( c("abcd", "defg"), split = "")
```

Remember, you can always select the element at a position in a list using double `[[]]`.  For example:
```{r}
strsplit("abcd", split = "")[[1]]
```

For `reverseComplement` you'll hopefully also be thinking that you'll want to use `gsub` to switch A -> T, G -> C and visa versa.

However, consider the following:

``` {r}
dna <- "ATGCATCG"
dna <- gsub("A","T", dna)
dna
dna <- gsub("G","C", dna)
dna
dna <- gsub("T","A", dna)
dna
dna <- gsub("C","G", dna)
dna
````

What happened here?  To get around this problem I like to use a little trick with cases: switch the case of the substitution letter then use the `toupper` function to convert lower case letters back to upper case.  For example:

``` {r}
dna <- "ATGCATCG"
dna <- gsub("A","t", dna)
dna
dna <- toupper(dna)
dna
````

As long as you've sanitized your inputs to this function and can rely on a starting case, it's a simple way to avoid overlapping substitutions.

## annotateFrame

This function is setup to accept a frame `offset` argument.  You'll call `annotateFrame` from annotate chromosome six times, once for each strand, and three times with `offset = 0`, `offset = 1`, and `offset = 2`.  If you're having trouble visualizing reading frames in sequence and traking starting positions I'd recommend drawing out a diagram for yourself.

If you call findORFs from annotateFrame remember that you'll want to convert lengths and startPositions from the amino acid coordinates to the correct DNA coordinates!

## loadFASTA

Take a look at the format of FASTA files.  You can find a truncated version of yeast chromosome 1 FASTA (DNA sequence) file here:  [data/chr01-truncated.fsa]("data/chr01-truncated.fsa").  Once we know our code is working we can run it on the full sequence, but for now it makes sense to use a smaller test dataset for troubleshooting.

Looking at the contents of this file, you'll probably want to omit the first line of this file because it doesn't contain sequence.  We can do that with indexing:

```{r}
dnaLines <- readLines("data/chr01-truncated.fsa")
length(dnaLines)

dnaLines <- dnaLines[2:length(dnaLines)]
```

Also we'll want to collapse all of the strings in this vector into a single string.  We can use paste to do that:

```{r}
dna <- paste(dnaLines, collapse = "", sep = "")
```

## Combining data.frame objects

In several places you'll might want to be able to combine `data.frame` tables, adding the rows of one table to another.  See the `merge`, `rbind` and `cbind` functions for a convenient way to do this.  Alternatively it can be done with the `tidyr` package; see [the cheatsheet](https://www.rstudio.com/wp-content/uploads/2015/02/data-wrangling-cheatsheet.pdf) for reference.

# Step 4: Annotate the whole genome!

Once you have your annotateChromosome function working with a sample chromosome FASTA file (takes the name of a chromosome FASTA file and produces a data.frame with ORF information), you're ready to write a function that annotates the entire yeast genome.

You can download the latest yeast genome sequence [here](http://downloads.yeastgenome.org/sequence/S288C_reference/chromosomes/fasta/).  Each of these files contains the DNA sequence for one of the yeast chromosomes.

This is probably a good time to introduce a Linux command line tool that is really useful when you need to pull files off the internet as part of a data analysis project.  In R studio, you can open a Linux command line shell, by opening Tools -> Shell.  The `wget` program takes will download a file from a URL to the local directory.

For example:

```SH
wget http://downloads.yeastgenome.org/sequence/S288C_reference/chromosomes/fasta/chr02.fsa
```

# Challenge #1:  Transcription factor binding site

The TATA-Binding Protein ([TBP](http://en.wikipedia.org/wiki/TATA-binding_protein)) is an important eukaryotic transcription factor.  It promotes transcription at many, but not all, eukaryotic genes.  It binds to the "TATA-box" sequence upstream of genes:

`TATAAA`

Modify your annotation script so that it produces a table listing TBP binding sites in the yeast genome.  Include a column that lists the likely ORF that is the target of TBP transcription target.  Remember that you'll have to keep track of both DNA strands.

What percentage of yeast genes are likely transcribed by TBP bound promoters?

# Challenge #2:  Handle introns

Yeast is a eukaryote, which means the protein coding sequences (ORFs) can be interrupted by introns.  Introns are defined by 3 sequences: start sequence, end sequence and a special sequence in the middle of the intron called a branch point sequence.  Unlike the TATA-Box however, each of these sequences comes in several different flavors.  Specifically, the variants are:

Start sites: 
```
GUAUGU 
GUAAGU 
GUAUGC 
GUAUGA 
GUACGU 
GUCAGU 
GUUAAG 
GUAGUA 
GCAUGU 
GUUCGU 
GUGAGU 
GCAAGU
```

Branch points:
```
GACUAAC 
UACUAAC 
AACUAAC 
AAUUAAC 
CACUAAC 
UGCUAAC 
UAUUAAC 
AGUUAAC 
CGUUAAC 
UGUUAAC 
CAUUAAC
```

End sites:
```
CAG
AAG
UAG
```

We can use regular expressions to find patterns that have variable characters at different positions.  Consider the following example:

``` {r}
grepl("bio297", "Hello bio297!")
grepl("bio297", "Hello bio397!")

#Match course number 297 or 397
grepl("bio[23]97", "Hello bio297!")
grepl("bio[23]97", "Hello bio297!")
````

See what happened there?  To match all splice end sites we could use the regular expression:

`"[CAU]AG"`

Re-factor your annotateChromosome function so that it first finds and then removes (hint: gsub) all of the likely introns from the chromosomal sequence before attempting to annotate the ORFs. You might 

How did your ORF list change when you accounted for introns?

